### ISSUE TITLE
Guroubi decides not to apply usercuts during the search?

### ISSUE THREAD

**Issue by CBongiova** [JuliaOpt/JuMP.jl#1391]
Hello, 

I have an MILP which I solve through a branch-and-cut procedure (hence using a callback function). I am printing out the violated valid inequalities and the number of explored nodes during the tree search. I have noticed that often times the console returns the same inequality for a different number of explored nodes, which leads me to suppose that the same usercut is being produced twice and at different nodes (hence affecting performance!). 

Is it possible that Gurobi didn't apply the usercut for a "low" violation but it is applying it later for a "higher" violation? 

For example, say the constraint is the following:
`@usercut(cb, sum(x[path[h],path[h+1]] for h=collect(1:length(path)-1)) <= (length(path)-3)`
Where path is a vector containing some node ids.
Then, using `node = MathProgBase.cbgetexplorednodes(cb)`,  the console prints out: 
> 
Num explored nodes 14.0: the constraint is produced for path [10, 13, 19, 42, 26, 34] with
 x_val 3.1328960560421835
Num explored nodes 15.0: the constraint is produced for path [10, 13, 19, 42, 26, 34] with x_val 3.14372170874214

**Comment by odow** [comment]
Hi @CBongiova,

A couple of comments:
 - it's a little hard to help without a [minimum working example](https://stackoverflow.com/help/mcve) that demonstrates the problem. Can you share a simple example?
- we've had too many problems like this with "solver-independent" callbacks. Hence, the new version of JuMP in development will require the user to implement solver-specific routines.

In future, please post questions like this to our [Discourse](https://discourse.julialang.org/c/domain/opt) site as a first port of call. It has many more people who read and might be able to help.

Thanks for using JuMP!

**Comment by CBongiova** [comment]
Hi @odow 

Thanks so much for your reply and suggestions (and preventively, for your time with the working example)
I have also posted the question to the Discourse site 👍 

Here is a working example:

`
using JuMP
using Gurobi
using MathProgBase
using Erdos
solver=GurobiSolver(PreCrush=1, Cuts=0, Presolve=0, Heuristics=0.0,TimeLimit=7200)

# Data Part
include("C:/Users/bongiova/Desktop/Scripts_Julia/EDARP/Runs_first_review/Runs1/readDAT_example.jl")
path_instance="C:/Users/bongiova/Desktop/Scripts_Julia/EDARP/Runs_first_review/Runs1/instances/a3-24-14.85-0.3-FULL.txt"
relax=false
numvehicles,numcustomers,numorigindepots,numdestdepots,numrecharging,numrepetitions,H,P,D,d,l,arr,dep,FO,FD,S,AOD,ADD,u,t,M_T=readDAT_example(path_instance)
numlocations=numcustomers*2+numorigindepots+numdestdepots+numrecharging+2*numvehicles

# Problem Sets
 A=collect(1:numlocations)
 K=collect(1:numvehicles)
 N=union(P,D)
 n=numcustomers

 m = Model(solver = solver)

 # Setting the decision variables
 @variable(m,x[i in A, j in A], Bin)
 @variable(m,T[i in A] >=0)

################
 # MODEL
 ################
# Setting the objective function
@objective(m, Min, sum(t[i,j]*x[i,j] for i = A for j = A))

@constraint(m, Const1[i=FO],sum(x[i,j] for j =AOD)==numvehicles)
@constraint(m, Const2[i=AOD],sum(x[i,j] for j =union(P,S,ADD))==1)
@constraint(m, Const3[j=ADD],sum(x[i,j] for i =union(D,S,AOD))==1)
@constraint(m, Const7[i=union(N,S)],sum(x[i,j] for j=setdiff(A,union(FO,FD)))-sum(x[j,i] for j=setdiff(A,union(FO,FD)))==0)
@constraint(m, Const7a[i=FO, j=FD],sum(x[i,k] for k=AOD)-sum(x[k,j] for k=ADD)==0)
@constraint(m, Const8[i=P],sum(x[i,j] for j=N if j!=i)==1)
@constraint(m, Const8b[j=D],sum(x[i,j] for i=N if j!=i)==1)
@constraint(m, Const6a[s=S],sum(x[s,i] for i=union(P,S,ADD))<=1)
@constraint(m, Const10[i=P],T[i]+d[i]+t[i,(i+n)]-T[(i+n)]<=0)
@constraint(m, Const11a[i=A], T[i]>=arr[i])
@constraint(m, Const11b[i=A], T[i]<=dep[i])
@constraint(m, Const12[i=P], T[(i+n)]-(T[i]+d[i])<=u[i])
@constraint(m, Const13[i=A,j=A;j!=i && M_T[i,j]>0], T[i]+t[i,j]+d[i]-M_T[i,j]*(1-x[i,j])<=T[j])
@constraint(m,Const26b[i=FD], T[i]<=H)

# Callback function
function mycutgenerator(cb)
    TOL=0.1
    path_length=7
X_values_paths=zeros(numcustomers,2*numcustomers+numrecharging)
traveltime_values_paths=zeros(numcustomers,2*numcustomers+numrecharging)
      for i=P
        temp=i
        X_val=0
        counter_its=1
        while counter_its<=path_length
          x_val=0.1
          next=0
          for j=union(N,S,ADD)
              x_val_new = getvalue(x[temp[end],j])
            if x_val_new>x_val
              x_val=x_val_new
              next=j
            end
         end #closes for j=union(N,S)

    if in(next,temp)==false && next!=0 && isempty(findin(ADD,next))
           X_val=X_val+x_val
           X_values_paths[i,counter_its]=x_val
           traveltime_values_paths[i,counter_its]=t[temp[end],next]+d[temp[end]]
           temp=hcat(temp,next)
           
           if length(temp)>2
               local_pickup=temp[1]
               idx_dropoff=findin(temp,local_pickup+numcustomers)
               if isempty(idx_dropoff)==false
               temp_local=temp[1:idx_dropoff[1]]
               X_val_infpath_ridetime=sum(X_values_paths[i,1:idx_dropoff[1]-1])
               traveltime_infpath_ridetime=sum(traveltime_values_paths[i,1:idx_dropoff[1]-1])

               if traveltime_infpath_ridetime-d[temp_local[end]]>u[temp_local[1]] && X_val_infpath_ridetime>(length(temp_local)-2)-1+TOL
                   node = MathProgBase.cbgetexplorednodes(cb)
                    println("Num explored nodes $node : the constraint is produced for path $temp_local with x_val $X_val_infpath_ridetime")
                    @usercut(cb, sum(x[temp_local[h],temp_local[h+1]] for h=collect(1:length(temp_local)-1)) <= (length(temp_local)-2)-1)
               end
         end #closes if isempty(idx_dropoff)==false
end #closes if length(temp)>2
end #closes the else statement checking for duplicates in the path/end of a path
         counter_its+=1
  end #closes the while
end #closes the loop for i
end #closes the callback function

# Tell JuMP/Gurobi to use our callback function
addcutcallback(m, mycutgenerator)
#solve
status=solve(m)
`
The data file: 
[a3-24-14.85-0.3-FULL.txt](https://github.com/JuliaOpt/JuMP.jl/files/2259359/a3-24-14.85-0.3-FULL.txt)

The .jl file used to read the data file (Here added as a .txt since I am not allowed to add .jl files). There is no need to review this code, you can just run it in the main code as this is just to read the data for the test.  
[readDAT_example.txt](https://github.com/JuliaOpt/JuMP.jl/files/2259360/readDAT_example.txt)

In this working example, I get: 

> 
Num explored nodes 597.0 : the constraint is produced for path [11, 45, 59, 14, 38, 18, 35] with x_v
al 5.416122094058215
Num explored nodes 597.0 : the constraint is produced for path [11, 45, 59, 14, 38, 18, 35] with x_v
al 5.416122094058215
... 
Num explored nodes 631.0 : the constraint is produced for path [4, 12, 23, 47, 28] with x_val 3.5145
12036318149
Num explored nodes 631.0 : the constraint is produced for path [4, 12, 23, 47, 28] with x_val 3.5145
12036318149
Num explored nodes 631.0 : the constraint is produced for path [4, 12, 23, 47, 28] with x_val 3.5145
12036318149
Num explored nodes 631.0 : the constraint is produced for path [4, 12, 23, 47, 28] with x_val 3.5145
12036318149

So it seems like the same constraint was produced multiple times? In this example the number of explored nodes remains the same (the constraint is added at the same node)? Is Gurobi looping over the callback function in the same B&B node?

**Comment by CBongiova** [comment]
Hi @odow 

Thanks so much for your reply and suggestions (and preventively, for your time with the working example)
I have also posted the question to the Discourse site 👍 

Here is a working example:

`using JuMP
using Gurobi
using MathProgBase
solver=GurobiSolver(PreCrush=1, Cuts=0, Presolve=0, Heuristics=0.0,TimeLimit=7200)

# Data Part
include("C:/Users/bongiova/Desktop/Scripts_Julia/EDARP/Runs_first_review/Runs1/readDAT_example.jl")
path_instance="C:/Users/bongiova/Desktop/Scripts_Julia/EDARP/Runs_first_review/Runs1/instances/a3-24-14.85-0.3-FULL.txt"
numvehicles,numcustomers,numorigindepots,numdestdepots,numrecharging,numrepetitions,H,P,D,d,l,arr,dep,FO,FD,S,AOD,ADD,u,t,M_T=readDAT_example(path_instance)
numlocations=numcustomers*2+numorigindepots+numdestdepots+numrecharging+2*numvehicles

# Problem Sets
 A=collect(1:numlocations)
 K=collect(1:numvehicles)
 N=union(P,D)
 n=numcustomers
 m = Model(solver = solver)

 # Setting the decision variables
 @variable(m,x[i in A, j in A], Bin)
 @variable(m,T[i in A] >=0)
 @variable(m,B[i in A] >=0)
 @variable(m,E[i in S] >=0)

################
 # MODEL
 ################
# Setting the objective function
@objective(m, Min, sum(t[i,j]*x[i,j] for i = A for j = A))

@constraint(m, Const1[i=FO],sum(x[i,j] for j =AOD)==numvehicles)
@constraint(m, Const2[i=AOD],sum(x[i,j] for j =union(P,S,ADD))==1)
@constraint(m, Const3[j=ADD],sum(x[i,j] for i =union(D,S,AOD))==1)
@constraint(m, Const7[i=union(N,S)],sum(x[i,j] for j=setdiff(A,union(FO,FD)))-sum(x[j,i] for j=setdiff(A,union(FO,FD)))==0)
@constraint(m, Const7a[i=FO, j=FD],sum(x[i,k] for k=AOD)-sum(x[k,j] for k=ADD)==0)
@constraint(m, Const8[i=P],sum(x[i,j] for j=N if j!=i)==1)
@constraint(m, Const8b[j=D],sum(x[i,j] for i=N if j!=i)==1)
@constraint(m, Const6a[s=S],sum(x[s,i] for i=union(P,S,ADD))<=1)
@constraint(m, Const10[i=P],T[i]+d[i]+t[i,(i+n)]-T[(i+n)]<=0)
@constraint(m, Const11a[i=A], T[i]>=arr[i])
@constraint(m, Const11b[i=A], T[i]<=dep[i])
@constraint(m, Const12[i=P], T[(i+n)]-(T[i]+d[i])<=u[i])
@constraint(m, Const13[i=A,j=A;j!=i && M_T[i,j]>0], T[i]+t[i,j]+d[i]-M_T[i,j]*(1-x[i,j])<=T[j])
@constraint(m,Const26b[i=FD], T[i]<=H)
@constraint(m,Const27[i=FO], B[i]==14.85)
@constraint(m, Const28[i=A],14.85>=B[i])
@constraint(m, Const29a[i=setdiff(A,S),j=setdiff(A,FD);j!=i],B[j]<=B[i]-0.055*t[i,j]+14.85*(1-x[i,j]))
@constraint(m, Const29b[i=setdiff(A,S),j=setdiff(A,FD);j!=i],B[j]>=B[i]-0.055*t[i,j]-14.85*(1-x[i,j]))
@constraint(m, Const29c[i=S,j=union(P,S,ADD);j!=i],B[j]<=B[i]+0.055*E[i]-0.055*t[i,j]+14.85*(1-x[i,j]))
@constraint(m, Const29d[i=S,j=union(P,S,ADD);j!=i],B[j]>=B[i]+0.055*E[i]-0.055*t[i,j]-14.85*(1-x[i,j]))
@constraint(m, Const30a[i=S,j=union(P,S,ADD);i!=j && M_T[i,j]>0],T[j]-t[i,j]-T[i]-M_T[i,j]*(1-x[i,j])<=E[i])
@constraint(m, Const30b[i=S,j=union(P,S,ADD);i!=j && M_T[i,j]>0],T[j]-t[i,j]-T[i]+M_T[i,j]*(1-x[i,j])>=E[i])

# Callback function
function mycutgenerator(cb)
    tt = time()
    TOL=0.1
    path_length=7

X_values_paths=zeros(numcustomers,2*numcustomers+numrecharging)
traveltime_values_paths=zeros(numcustomers,2*numcustomers+numrecharging)

      for i=P
        temp=i
        X_val=0
        counter_its=1
        capacity_path=1
        while counter_its<=path_length
          check=false
          x_val=0.1
          next=0
          for j=union(N,S,ADD)
              x_val_new = getvalue(x[temp[end],j])
            if x_val_new>x_val
              x_val=x_val_new
              next=j
            end
         end #closes for j=union(N,S)

    if in(next,temp)==false && next!=0 && isempty(findin(ADD,next))
           X_val=X_val+x_val
           X_values_paths[i,counter_its]=x_val
           traveltime_values_paths[i,counter_its]=t[temp[end],next]+d[temp[end]]
           temp=hcat(temp,next)
          
           if length(temp)>2
               local_pickup=temp[1]
               idx_dropoff=findin(temp,local_pickup+numcustomers)
               if isempty(idx_dropoff)==false
               temp_local=temp[1:idx_dropoff[1]]
               X_val_infpath_ridetime=sum(X_values_paths[i,1:idx_dropoff[1]-1])
               traveltime_infpath_ridetime=sum(traveltime_values_paths[i,1:idx_dropoff[1]-1])

               if traveltime_infpath_ridetime-d[temp_local[end]]>u[temp_local[1]] && X_val_infpath_ridetime>(length(temp_local)-2)-1+TOL
                   node = MathProgBase.cbgetexplorednodes(cb)
                    println("Num explored nodes $node : the constraint is produced for path $temp_local with x_val $X_val_infpath_ridetime")
                    @usercut(cb, sum(x[temp_local[h],temp_local[h+1]] for h=collect(1:length(temp_local)-1)) <= (length(temp_local)-2)-1)
                    check=true
               end
         end #closes if isempty(idx_dropoff)==false
end #closes if length(temp)>2

if check==true
    break
end

      elseif in(next,temp)==true && next!=0
                counter_its=1000
         elseif !isempty(findin(ADD,next))
               counter_its=1000
         elseif next==0
               counter_its=1000

         end #closes the else statement checking for duplicates in the path/end of a path
         counter_its+=1

  end #closes the while
end #closes the loop for i
end #closes the callback function

# Tell JuMP/Gurobi to use our callback function
addcutcallback(m, mycutgenerator)
#solve
status=solve(m)
`

The data file: 
[a3-24-14.85-0.3-FULL.txt](https://github.com/JuliaOpt/JuMP.jl/files/2259359/a3-24-14.85-0.3-FULL.txt)

The .jl file used to read the data file (Here added as a .txt since I am not allowed to add .jl files). There is no need to review this code, you can just run it in the main code as this is just to read the data for the test.  
[readDAT_example.txt](https://github.com/JuliaOpt/JuMP.jl/files/2259360/readDAT_example.txt)

In this working example, I get (for example): 
> Num explored nodes 315.0 : the constraint is produced for path [4, 47, 12, 29, 36, 28] with x_val 4.66579735844086
Num explored nodes 316.0 : the constraint is produced for path [4, 47, 12, 29, 36, 28] with x_val 4.969837497678187

So it seems like the same constraint was produced multiple times?  Is it possible that Gurobi loops over the callback function in the same B&B node while other nodes are being explored in parallel?

**Comment by odow** [comment]
> I have also posted the question to the Discourse site 

You should edit the discourse post to link to this issue. (Always do this when posting the same question in multiple places.)

Some advice crafting a working example. One of the biggest things to do is make it minimal. From the [stackoverflow help](https://stackoverflow.com/help/mcve):
> Minimal – Use as little code as possible that still produces the same problem

The less code people have to wade through, the easier it is to help. I didn't run your code because I don't have Erdos installed, there are lot of files, and there is a lot of things going on. Try to remove as much code as possible so that it still demonstrates the problem.

**Question**: do you mean to add a [lazy constraint](http://www.gurobi.com/documentation/8.0/refman/c_grbcblazy.html) or a [user cut](http://www.gurobi.com/documentation/8.0/refman/c_grbcbcut.html)?


> is there a way to actually print the current b&b node instead of the number of explored nodes in the callback?

Not to my knowledge. It is one of the reasons why we are removing solver independent callbacks as they often require solver-specific information.

> I think one of the major attractives of JuMP (and Julia for the people in OR) is indeed that the syntax does not change among solvers

The problem is that callbacks are tied to particular solvers. The syntax obscures a lot of problems from users. (See, for example, https://github.com/JuliaOpt/GLPKMathProgInterface.jl/issues/19.)

> does this mean that I will have to update the syntax of all my callbacks in near future?

Yes.

> Will it remain easy to learn and implement?

You will have to understand the particular quirks of the solver you are using and how it implements callbacks. [Here](https://github.com/JuliaOpt/Gurobi.jl/blob/0d73b3e35b9fac7e25eb0d4f4776a76d53deefd3/test/MOIWrapper.jl#L126-L181) is an example of lazy cuts in Gurobi with MOI (not JuMP). For example, you need to know that you can only add lazy constraints when called from `Gurobi.CB_MIPSOL`. There will be some nice helper functions from JuMP, but I am unsure what form these will take at present.

**Comment by CBongiova** [comment]
Hi @odow, 

> You should edit the discourse post to link to this issue. (Always do this when posting the same question in multiple places.)

Thanks I have added the link to this post on the discourse 

> Some advice crafting a working example. One of the biggest things to do is make it minimal. From the stackoverflow help:
Minimal – Use as little code as possible that still produces the same problem
The less code people have to wade through, the easier it is to help. I didn't run your code because I don't have Erdos installed, there are lot of files, and there is a lot of things going on. Try to remove as much code as possible so that it still demonstrates the problem.

You can remove the line "using Erdos". Actually you won't need it in the working example. Sorry I forgot to remove it myself. 

I have started from a minimal MILP model and kept adding decision variables and constraints until I have encountered the same issue I am experiencing right now. Further reducing the model makes the problem so easy to solve that the issue is not replicated. As for the callback function, I have only kept one valid inequality (which is the one being replicated). 

The issue is also data-dependent: With different parameters, the solver would possibly not behave the same way. Reproducing another dataset which creates the same issue in Gurobi would probably be very complicated to do. That's why I have attached a data file and a simple script to read it. 

The code is actually quite simple to be used: Download the attached files, change the path to your local path in the working_example code -- lines 7 and 8, and run. The working example is actually a quite compact MILP model and a very small callback function (reproducing only the user cuts which create the issue) 

> Question:do you mean to add a lazy constraint or a user cut?
I mean to add a user cut because the constraint does not define the problem but should rather cut an infeasible fractional solution during the search. The problem is that if Gurobi keeps adding the same valid inequality during the search, the model becomes heavier and therefore CPU time increases. This is something that I would like to avoid. 

Thank you for all the other information! 
I really hope someone will be able to help me on this issue... :(

**Comment by mlubin** [comment]
Closing since the discussion ended and there's no evidence of a JuMP issue. Feel free to reopen or create a new issue if appropriate. As far as I can tell, the issue is about behavior internal to Gurobi.
